aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -3,48 +3,117 @@
 #include "ATen/ExpandUtils.h"
 #include "ATen/NativeFunctions.h"
 #include "ATen/WrapDimUtils.h"
-#include "cpu/UnaryOpsKernel.h"
+
+#include "ATen/CPUApplyUtils.h"
+#include "ATen/Parallel.h"
+#include "ATen/native/cpu/UnaryOpsKernel.h"
 
 #include <algorithm>
+#include <cmath>
 #include <functional>
+#include <iostream>
 #include <numeric>
 #include <vector>
 
 #include <map>
 
 namespace at { namespace native {
 
-#define IMPLEMENT_UNARY_OP(op)                                                \
-Tensor op(const Tensor& self) {                                               \
-  Tensor result = self.type().tensor();                                       \
-  return at::op ## _out(result, self);                                        \
-}                                                                             \
-Tensor& op##_(Tensor& self) {                                                 \
-  return at::op ## _out(self, self);                                          \
-}                                                                             \
-Tensor& _ ## op ## _out_cuda(Tensor& result, const Tensor& self) {            \
-  return at::_ ## op ## _out(result, self);                                   \
-}                                                                             \
-Tensor& _ ## op ## _out_cpu(Tensor& result, const Tensor& self) {             \
-  if (result.is_contiguous() && self.is_contiguous()) {                       \
-    result.resize_(self.sizes());                                             \
-    if (result.numel() > 0) {                                                 \
-      op ## Impl(result, self);                                               \
-    }                                                                         \
-    return result;                                                            \
-  }                                                                           \
-  return at::_ ## op ## _out(result, self);                                   \
+static Tensor _sort_strides(Tensor& tensor_) {
+  std::vector<int64_t> strides = tensor_.strides();
+  std::vector<int64_t> indices(strides.size());
+  iota(indices.begin(), indices.end(), 0);
```

Comment:
This could be made slightly faster. Currently it will zero fill `indices` only to overwrite it a moment later. We could use `reserve` and `push_back` consecutive integers

Reply:
[iota](http://en.cppreference.com/w/cpp/algorithm/iota) fills it with a sequence (0, 1, 2, 3....) which is necessary for the sort.